File: Services/VkService/src/main/java/services/VkService.java
package services;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.sun.istack.NotNull;
import datamodels.AppProperties;
import datamodels.ImageLink;
import datamodels.WorkData;
import entities.VkGroupEntity;
import entities.VkNewsEntity;
import org.hibernate.Criteria;
import org.hibernate.SQLQuery;
import org.hibernate.Session;
import org.hibernate.criterion.Restrictions;
import utils.HibernateSessionFactory;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;


public class VkService {

    public VkService(AppProperties appProperties) {
        AppProperties appProperties1 = appProperties;
    }

    public void saveNews(WorkData workData, Session session) {
        List<VkNewsEntity> data = convertData(workData, session);
        saveData(data, session);
    }

    private List<VkNewsEntity> convertData(WorkData workData, Session session) {
        List<Integer> groupIds = (List<Integer>) workData.getGroupIds();
        List<Integer> newsIds = (List<Integer>) workData.getNewsIds();
        Collections.sort(newsIds);

        List<VkNewsEntity> result = new ArrayList<>(workData.getNews().size());
        List<Integer> updateGroups = new ArrayList<>(groupIds.size());
        long userId = workData.getUserId();

        // get new only and convert
        for (JsonElement item : workData.getNews()) {
            JsonObject json = item.getAsJsonObject();

            if (json.get("marked_as_ads").getAsInt() == 1) {
                continue;
            }

            ImageLink imageLink = getImageLink(json);
            Integer postId = json.get("post_id").getAsInt();
            Integer groupId = -json.get("source_id").getAsInt();

            if (Collections.binarySearch(newsIds, postId) >= 0) {
                continue;
            }

            if (!groupIds.contains(groupId)) {
                // add new group for user
                addGroup(groupId, workData, session);
                groupIds.add(groupId);
            } else {
                updateGroups.add(groupId);
            }

            VkNewsEntity entity = new VkNewsEntity();
            entity.setGroupId(groupId);
            entity.setPostId(postId);
            entity.setText(json.get("text").getAsString());
            entity.setUserId(userId);
            entity.setImage(imageLink.getImage());
            entity.setLink(imageLink.getLink());
            entity.setTimestamp(json.get("date").getAsLong());

            result.add(entity);
        }

        //cleanGroups(workData, groupIds, session);  // todo: in thread
        updateGroups(workData, groupIds, session);

        return result;
    }

    private void updateGroups(WorkData workData, List<Integer> groupIds, Session session) {
        session.beginTransaction();
        Criteria groupCriteria = session.createCriteria(VkGroupEntity.class);
        groupCriteria.add(Restrictions.in("id", groupIds));
        List<VkGroupEntity> entities = groupCriteria.list();

        for (VkGroupEntity entity : entities) {
            for (JsonElement item : workData.getGroups()) {
                JsonObject json = item.getAsJsonObject();
                int id = json.get("gid").getAsInt();

                if (entity.getGid() == id) {
                    entity.setName(json.get("name").getAsString());
                    entity.setLinkedName(json.get("screen_name").getAsString());
                    entity.setImage(json.get("photo").getAsString());

                    try {
                        session.save(entity);
                    } catch (Exception e) {
                        continue;
                    }
                }
            }
        }

        try {
            session.getTransaction().commit();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void saveData(List<VkNewsEntity> news, Session session) {
        session.beginTransaction();
        SQLQuery query = session.createSQLQuery("SET NAMES utf8mb4");
        query.executeUpdate();

        for (VkNewsEntity item : news) {
            try {
                session.save(item);
            } catch (Exception e) {
                continue;
            }
        }

        try {
            session.getTransaction().commit();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void addGroup(int groupId, WorkData workData, Session session) {
        for (JsonElement item : workData.getGroups()) {
            JsonObject json = item.getAsJsonObject();
            int id = json.get("gid").getAsInt();

            if (groupId == id) {
                String name = json.get("name").getAsString();
                String linkedName = json.get("screen_name").getAsString();
                String image = json.get("photo").getAsString();

                session.beginTransaction();
                VkGroupEntity groupEntity = new VkGroupEntity(groupId, workData.getUserId(), name, linkedName, image);
                session.save(groupEntity);
                session.getTransaction().commit();

                return;
            }
        }
    }

    // delete group from DB if any group not in JSON
    private void cleanGroups(WorkData workData, List<Integer> groupIds, Session session) {

        // get all grous for user
        Criteria criteria = session.createCriteria(VkGroupEntity.class);
        criteria.add(Restrictions.eq("userId", workData.getUserId()));

        List<VkGroupEntity> groups = (List<VkGroupEntity>) criteria.list();
        List<VkGroupEntity> oldGroups = new ArrayList<>();

        // get old groups
        for (VkGroupEntity group : groups) {
            if (!groupIds.contains(group.getGid())) {
                oldGroups.add(group);
            }
        }

        if (oldGroups.size() == 0) {
            return;
        }

        // delete all old groups if it is not in groupIds
        session.beginTransaction();

        for (VkGroupEntity vkGroupEntity : oldGroups) {
            session.delete(vkGroupEntity);
        }

        session.getTransaction().commit();
    }

    @NotNull
    private ImageLink getImageLink(JsonObject json) {
        if (!json.has("attachment")) {
            return new ImageLink("", "");
        }


        JsonObject attachment = json.get("attachment").getAsJsonObject();
        String postType = attachment.get("type").getAsString();
        String image = null;
        String link = null;

        if (postType.equals("photo")) {
            if (attachment.has("photo")) {
                JsonObject photo = attachment.getAsJsonObject("photo");

                if (photo.has("src_big")){
                    image = photo.get("src_big").getAsString();
                }
            }
        } else if (postType.equals("link")) {
            JsonObject linkObject = attachment.getAsJsonObject("link");

            if (linkObject.has("image_big")) {
                image = linkObject.get("image_big").getAsString();
            } else if (linkObject.has("image_src")) {
                image = linkObject.get("image_src").getAsString();
            }

            link = linkObject.get("url").getAsString();
        } else if (postType.equals("doc")) {
            JsonObject doc = attachment.get("doc").getAsJsonObject();

            if (doc.has("ext") && doc.get("ext").getAsString().equals("gif")) {
                image = doc.get("url").getAsString();
            }
        }

        return new ImageLink(image, link);
    }

}

File: Services/vk_service/src/vk_service.rs
use serde_json::map::Map;
use serde_json::Value;
use mysql;
use std::collections::HashSet;

use model::json_models::AppSettings;
use model::db_models::VkGroup;
use model::db_models::VkNews;
use model::data_models::ImageLink;


pub struct VkService<'a> {
    settings: &'a AppSettings,
}

impl<'a> VkService<'a> {

    pub fn new(settings: &'a AppSettings) -> VkService {
        VkService { settings }
    }

    pub fn update(&self, json: Map<String, Value>, user_id: u64) {
        let _data = self.convert_data(json, user_id);
    }

    fn get_existing_group_ids(&self, user_id: u64) -> Vec<u64> {
        let pool = self.get_pool();
        let mut group_ids: Vec<u64> = pool.prep_exec("select Gid from vkgroups where UserId = :user_id", params!{"user_id" => user_id})
            .map(|result| {
                result.map(|x| x.unwrap()).map(|row| {
                    let group_id: u64 = mysql::from_row(row);

                    return group_id
                }).collect()
            }).unwrap();
        group_ids.sort();

        return group_ids;
    }

    fn get_existing_news_ids(&self, user_id: u64) -> Vec<u64> {
        let pool = self.get_pool();
        let mut news_ids: Vec<u64> = pool.prep_exec("select PostId from vknews where UserId = :user_id", params!{"user_id" => user_id})
            .map(|result| {
                result.map(|x| x.unwrap()).map(|row| {
                    let news_id: u64 = mysql::from_row(row);

                    return news_id
                }).collect()
            }).unwrap();
        news_ids.sort();

        return news_ids;
    }

    fn convert_data(&self, json: Map<String, Value>, user_id: u64) -> Vec<VkNews> {
        let news = match json["items"].as_array() {
            Some(arr) => arr,
            None => return vec![],
        };
        let mut group_ids = self.get_existing_group_ids(user_id);
        let news_ids = self.get_existing_news_ids(user_id);
        let groups = self.get_groups(&json);
        let mut result = vec![];

        println!("{:?}", news.len());

        for item in news {
            let news_json = match item.as_object() {
                Some(obj) => obj,
                None => continue,
            };
            if let Some(i) = news_json["marked_as_ads"].as_u64() {
                if i == 1 {
                    continue;
                }
            }
            let post_id = match news_json["post_id"].as_u64() {
                Some(id) => id,
                None => continue,
            };
            let group_id = match news_json["source_id"].as_i64() {
                Some(id) => (id - (id * 2)) as u64,
                None => continue,
            };
            let mut update_groups = vec![];

            if let Ok(_) = news_ids.binary_search(&post_id) {
                continue;
            }

            match group_ids.binary_search(&group_id) {
                Err(_) => {
                    group_ids.push(group_id);
                    group_ids.sort();
                    self.add_group(group_id, &groups);
                },
                Ok(_) => {
                    update_groups.push(group_id);
                }
            }

            let image_link = self.get_image_link(news_json);
            let entity = VkNews{
                id: post_id,
                user_id: user_id,
                group_id: group_id,
                post_id: post_id,
                text: news_json["text"].as_str().unwrap().to_owned(),
                timestamp: news_json["date"].as_u64().unwrap(),
                image: image_link.image,
                link: image_link.link
            };
            result.push(entity);
        }

        self.update_groups(group_ids, &groups);

        return result;
    }

    fn update_groups(&self, group_ids: Vec<u64>, groups: &Vec<VkGroup>) {
        println!("HHHHHHHHHHHHHHHHHHHHHHHHHHHHH");
        // get db groups
        let pool = self.get_pool();
        let entities: Vec<VkGroup> = pool.prep_exec("select Id, Gid, UserId, Name, LinkedName, Image from vkgroups where Gid in (:ids)", (group_ids))
            .map(|result| {
                result.map(|x| x.unwrap()).map(|row| {
                    let (id, gid, user_id, name, linked_name, image) = mysql::from_row(row);

                    return VkGroup {
                        id: id,
                        gid: gid,
                        user_id: user_id,
                        name: name,
                        linked_name: linked_name,
                        image: image,
                    }
                }).collect()
            }).unwrap();

            println!("{:?}", entities);

        let mut stmt = pool.prepare("update vkgroups set Name = :name, LinkedName = :screen_name, Image = :photo where Gid = :gid").unwrap();

        for entity in entities {
            if let Some(group) = groups.iter().find(|&x| x.gid == entity.gid) {
                let data = params!{
                    "name" => &group.name,
                    "screen_name" => &group.linked_name,
                    "photo" => &group.image,
                };
                // stmt.execute(data).unwrap_or_else(|err| println!("updat{}", group.gid));
                stmt.execute(data).unwrap();
            }
        }
    }

    fn get_groups(&self, json: &Map<String, Value>) -> Vec<VkGroup> {
        let mut groups = vec![];
        let groups_json = match json["groups"].as_array() {
            Some(obj) => obj,
            None => return groups,
        };

        for item in groups_json {
            let mut group = VkGroup::default();

            if let Some(gid) = item["gid"].as_u64() {
                group.gid = gid;
            };
            if let Some(name) = item["name"].as_str() {
                group.name = name.to_owned();
            };
            if let Some(linked_name) = item["screen_name"].as_str() {
                group.linked_name = linked_name.to_owned();
            };
            if let Some(image) = item["photo"].as_str() {
                group.image = image.to_owned();
            };

            groups.push(group);
        }

        return groups;
    }

    fn add_group(&self, group_id: u64, groups: &Vec<VkGroup>) {
        if let Some(group) = groups.iter().find(|&x| x.gid == group_id) {
            let pool = self.get_pool();
            let mut stmt = pool.prepare("insert into vkgroups (gid, UserId, Name, LinkedName, Image) values (:gid, :user_id, :name, :linked_name, :image)").unwrap();
            stmt.execute(params!{
                "gid" => group.gid,
                "user_id" => group.user_id,
                "name" => &group.name,
                "linked_name" => &group.linked_name,
                "image" => &group.image,
            }).unwrap();
        }
    }

    fn get_image_link(&self, json: &Map<String, Value>) -> ImageLink {
        if !json.contains_key("attachment") {
            return ImageLink::default();
        }

        let attachment = match json["attachment"].as_object() {
            Some(obj) => obj,
            None => return ImageLink::default(),
        };
        let post_type = match attachment["type"].as_str() {
            Some(string) => string,
            None => return ImageLink::default(),
        };
        let mut image = String::new();
        let mut link = String::new();

        if post_type == "photo" {
            if attachment.contains_key("photo") {
                let photo = match attachment["photo"].as_object() {
                    Some(obj) => obj,
                    None => return ImageLink::default(),
                };
                if photo.contains_key("src_big") {
                    image = match photo["src_big"].as_str() {
                        Some(string) => string.to_owned(),
                        None => return ImageLink::default(),
                    };
                }
            }
        } else if post_type == "link" {
            let link_object = match attachment["link"].as_object() {
                Some(obj) => obj,
                None => return ImageLink::default(),
            };
            if link_object.contains_key("image_big") {
                image = match link_object["image_big"].as_str() {
                    Some(string) => string.to_owned(),
                    None => return ImageLink::default(),
                };
            } else if link_object.contains_key("image_src") {
                image = match link_object["image_src"].as_str() {
                    Some(string) => string.to_owned(),
                    None => return ImageLink::default(),
                };
            }
            link = match link_object["url"].as_str() {
                Some(string) => string.to_owned(),
                None => return ImageLink::default(),
            };
        } else if post_type == "doc" {
            let doc = match attachment["doc"].as_object() {
                Some(obj) => obj,
                None => return ImageLink::default(),
            };
            if doc.contains_key("ext") {
                let ext = match doc["ext"].as_str() {
                    Some(string) => string,
                    None => return ImageLink::default(),
                };
                if ext == "gif" {
                    image = match doc["url"].as_str() {
                        Some(string) => string.to_owned(),
                        None => return ImageLink::default(),
                    };
                }
            }
        }

        return ImageLink::new(image, link);
    }

    fn get_pool(&self) -> mysql::Pool {
        let mut builder = mysql::OptsBuilder::new();
        builder.ip_or_hostname(Some(self.settings.db_host.clone()))
            .db_name(Some(self.settings.db_name.clone()))
            .pass(Some(self.settings.db_password.clone()))
            .user(Some(self.settings.db_user.clone()))
            .prefer_socket(false);

        return mysql::Pool::new(builder).unwrap();
    }

}